pkg/reconciler/clusteringress/resources/virtual_service.go

```diff
@@ -59,24 +65,54 @@ func MakeVirtualService(ci *v1alpha1.ClusterIngress, gateways []string) *v1alpha
 	ingressLabels := ci.Labels
 	vs.Labels[serving.RouteLabelKey] = ingressLabels[serving.RouteLabelKey]
 	vs.Labels[serving.RouteNamespaceLabelKey] = ingressLabels[serving.RouteNamespaceLabelKey]
+	return vs
+}
 
+// MakeMeshVirtualService creates Istio VirtualService as network
+// programming for Istio network mesh.
+func MakeMeshVirtualService(ci *v1alpha1.ClusterIngress) *v1alpha3.VirtualService {
+	vs := &v1alpha3.VirtualService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            names.MeshVirtualService(ci),
+			Namespace:       VirtualServiceNamespace(ci),
+			OwnerReferences: []metav1.OwnerReference{*kmeta.NewControllerRef(ci)},
+			Annotations:     ci.ObjectMeta.Annotations,
+		},
+		Spec: *makeVirtualServiceSpec(ci, []string{"mesh"}, retainLocals(getHosts(ci))),
```

Comment:
And `[]string{"mesh"}` can be initialize once `sets.Strings`. Then if we can make `gateways` also a set probably above in the call stack, then we can remove the `intersect` method altogether.

Reply:
We won't intersect with `gateways`, I think you mean `hosts`.  As with the other request, I will follow up since it result in a large test change and may also result in potential large change during upgrade.